Measure runtime sigops / sighashbytes requirements

dgenr8 opened the pull request:
Plumbing to measure two resources required to validate a block or transaction: runtime signature check operation count, and the number of bytes hashed at runtime for signature checks.

To set consensus block limits based on these measurements, pass the limits when initializing the ValidationResourceTracker in ConnectBlock.  CheckInputs will fail if they are exceeded.

To access transaction measurements in CreateNewBlock, use the new GetRuntimeSigOpCount() and GetRuntimeSighashBytes() measurements in CTxMempoolEntry.

zander commented:
Classic shipped this for a while and has done various tests.
Since you seem to be curious I'll share the findings :)
First of all, the most common critique was that we replace one hard limit with another. There in practice is very very little difference in measuring one over the other. You can still attack in either case and the amount of damage done is minimum to none for modern miner hardware.

There are two solutions that I feel are the way to go for fixing this properly;

* First of all, we remove the O(n^2) by introducing Flexible Transactions. This means we no longer need the check in the first place. But this is an invasive hard fork and we need a lot of time before all people have upgraded to use it.
* much easier is introducing optimistic mining. This means that a slow-to-validate block will be orphaned immediately because there is a timeout how long a miner is going to wait before a block will be used to mine on top of.

The second has the advantage that it can be rolled out without any changes to the consensus rules. Thus we don't need to wait for another hard or soft fork, we can fix it when its ready.

I'm sure there are other solutions in the works, the PV concept is promising as well although its currently untested.

And, to help you guys out with even more info, here is a spreadsheet on the amount of sigops we *actually* saw over the last year in Bitcoin.

[sigops.zip](https://github.com/BitcoinUnlimited/BitcoinUnlimited/files/545722/sigops.zip)

(apologies for the zip file, github doesn't like other filetypes)..

dgenr8 commented:
A validation timeout proportional to EB or actual blocksize would be similar to a proportional limit on bytes hashed, but would be dependent on local hardware performance.

ftrader commented:
Given what dgenr8 has pointed out, I don't quite see how a timeout is going to help enforce consensus except under some very strong assumptions (which may be justifiable, but they have to be spelled out!)

I much prefer fixed constraints on sigops, even if not relaxed further, until a hard fork to fixing the O(n^2) can be rolled out.

zander commented:
> I much prefer fixed constraints on sigops

I think that is a good idea too.

ftrader commented:
I would also submit that adding any timeouts based on wallclock time should be avoided wherever possible.
They are expensive to test (because slow), they make testing itself brittle (tests failing because some machines have different ideas of what the time is and how fast it should pass). Finally, they add yet more magic numbers to the code which needs constant maintenance (since they depend on changing external factors like processing speeds, latencies). Yuck.

deadalnix commented:
I don't think we should go the sighashbytes direction. We could measure everything, but in the end, it is better to try to make the hashing as fast as possible rather than measure how slow it is and try to work around it. plus this limit doesn't exist outside of classic yet, and I don't think it should become a standard.

deadalnix commented:
To clarify after talking with @dgenr8 , I'm not against this PR, but I don't think it should include the part where it counts the byte hashed. That is only useful if the plan to create a consensus rule around byte hashed, which I think isn't what we want.

deadalnix commented:
@dgenr8 You haven't updated that one in a while, Do you ant me to take over that one ?

dgenr8 committed:
Allow precise tracking of validation sigops/bytes

Adds a ValidationResourceTracker class that is passed to
CheckInputs() / CScriptCheck() to keep track of the exact number
of signature operations required to validate something, and the
exact number of bytes hashed to compute signature hashes.

Also extends CHashWriter to keep track of number of bytes hashed.

This is a subset of the original work by Gavin Andresen.

dgenr8 commented:
These statistics are used only for logging in #164, so the ability to set limits in a ValidationResourceTracker, and more importantly the consensus checks against these limits, have been removed.

The mempool tracking of runtime resource usage has also been removed (it was intended for use in CreateNewBlock).
